eth/beacon/state_machines/base.py

```diff
@@ -67,10 +71,16 @@ def import_block(self, block: BaseBeaconBlock) -> Tuple[BeaconState, BaseBeaconB
 class BeaconStateMachine(BaseBeaconStateMachine):
     def __init__(self,
                  chaindb: BaseBeaconChainDB,
-                 block: BaseBeaconBlock,
-                 state: BeaconState) -> None:
-        # TODO: get state from DB, now it's just a stub!
-        self.state = state
+                 block_root: Hash32) -> None:
+        self.chaindb = chaindb
+        self.chaindb.set_block_class(self.get_block_class())
```

Comment:
First, I'm worried about the statefulness this introduces to the `chaindb` instance which is a shared resource.  This code is very likely to become problematic within the `trinity` context since there will be multiple processes/coroutines interacting with the `chaindb` and thus if this value changes mid-stream we'll get weird bugs.

I'd like to explore how we can get away from this pattern.  

Properties we need from the API:

- the `chaindb` *needs* to know the proper `BlockClass` to use.  
- Ideally we avoid having to specify a new `ChainDB` class for every fork since it's likely that the internal logic won't need to be changed.

And last, we've been wanting to move away from the inheritance based model that we are currently using in the current `eth.vm` namespace in favor of something that is composition based.

> Forgive 

### Idea A: pass `block_class` as a parameter into the `ChainDB.__init__` class.

* Benefits:
  - compositional
* Cons
  - requires the `chaindb` to exist at the fork level rather than the chain level, meaning it will need to be re-instantiated for each fork.  This may result in some larger refactors.

### Idea B: pass `block_class` into any method that needs it.

* Benefits
  - makes the `ChainDB` agnostic about the rich encoding classes.
* Cons
  - requires much more explicit code

### Idea C: create an `Encoder` class which acts as the translator between the `ChainDB` and the `StateMachine`.  Each `StateMachine` would require an `Encoder` class like this:

```python
BlockClass = TypeVar("BlockClass", bound=BaseBlock)

class Encoder(Generic[BlockClass]):
    @classmethod
    def get_block_class(cls) -> BlockClass:
        ...
    def encode_block(block: BlockClass) -> bytes:
        ...
    def decode_block(raw_block: bytes) -> BlockClass:
        ...
```

Then each `StateMachine` would need to be told which `Encoder` class to use which ideally isn't done with inheritance but a class variable seems like the ideal way to do this...

And then the `ChainDB` would need to be given the `Encoder` instance via some API, potentially by passing it into any method that needs it.

### Closing thoughts

I think I'm leaning towards either B or C, with both being different approaches to the same concept.  I think passing in the data to `chaindb.__init__` may prove problematic as it would move the `ChainDB` from being a `Chain` level object down to being a `StateMachine` level object, thus I think it may be appropriate to require each method which needs this data to require it as a parameter.  Note that option C allows a smooth transition from `RLP` to `SSZ` since the `Encoder` class abstracts away any encoding details.